### PR TITLE
Fixes #28966 - disallow save project in postgres when < 9.5

### LINKED ISSUE #28966: Save project to PostGreSQL doesn't work with PostGreSQL 9.4 and older
Author Name: **Michel Stuyts** (@mstuyts)
Original Redmine Issue: [21148](https://issues.qgis.org/issues/21148)
Affected QGIS version: 3.4.4
Redmine category:project_loading/saving

---
The functionality to save a project to PostgreSQL will only work with PostgreSQL versions 9.5 and newer, because:
* the datatype JSONB was introduced in PostgreSQL 9.4
* the ON CONFLICT clause was introduced in  PostgreSQL 9.5

Problem:
* It's possible to check "Allow saving/loading QGIS projects in the database" in the Database Connection Settings for PostgreSQL databases older than version 9.5.
* When this checkbox is checked, it is possible to select Save To -> PostgreSQL in the Menu. After selecting a database, a schema and a project it will show an error that doesn't show the problem is the PostgreSQL version that is being used.

Expected behavior:
* It's not possible to check the checkbox "Allow saving/loading QGIS projects in the database" for PostgreSQL databases older than version 9.5. 
* Somewhere should be mentioned that this functionality only works with version 9.5 and newer.

Ideal behavior:
* Saving projects to PostgreSQL also works in older versions of PostgreSQL

### PR DESCRIPTION
**Pull request by stev-0**
## Description
Disables save project in database where Postgres database is version 9.4 or less. Relies on testing connection to determine postgres version. 

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/providers/postgres/qgspgnewconnection.cpp [hunk 1]
```diff
@@ -205,6 +206,18 @@ void QgsPgNewConnection::testConnection()
...
+  if ( conn->postgresVersion() < 95000 )
+  {
+    cb_projectsInDatabase->setChecked( false );
+    cb_projectsInDatabase->setEnabled( false );
+    cb_projectsInDatabase->setToolTip( QStringLiteral( "Saving projects in databases not available for Postgres databases earlier than 9.5" ) );
+  }
+  else
+  {
+    cb_projectsInDatabase->setEnabled( true );
+    cb_projectsInDatabase->setToolTip( QStringLiteral( "" ) );
```
Comment (nyalldawson) [inline src/providers/postgres/qgspgnewconnection.cpp]:
```suggestion
    cb_projectsInDatabase->setToolTip( QString() );
```

**Comment by jef-n** [comment]
Why does that need to be disabled?

**Comment by stev-0** [comment]
See https://github.com/qgis/QGIS/issues/28966

File: src/providers/postgres/qgspostgresconn.cpp [hunk 2]
```diff
@@ -871,6 +871,11 @@ bool QgsPostgresConn::hasPointcloud()
   return mPointcloudAvailable;
 }
 
+int QgsPostgresConn::postgresVersion()
```
Comment (jef-n) [inline src/providers/postgres/qgspostgresconn.cpp]: duplicates QgsPostgresConn::pgVersion()

File: src/providers/postgres/qgspgnewconnection.cpp [hunk 3]
```diff
@@ -205,6 +206,18 @@ void QgsPgNewConnection::testConnection()
 
   QgsPostgresConn *conn = QgsPostgresConn::connectDb( uri.connectionInfo( false ), true );
 
+  if ( conn->pgVersion() < 90500 )
+  {
+    cb_projectsInDatabase->setChecked( false );
+    cb_projectsInDatabase->setEnabled( false );
+    cb_projectsInDatabase->setToolTip( QStringLiteral( "Saving projects in databases not available for Postgres databases earlier than 9.5" ) );
```
Comment (m-kuhn) [inline src/providers/postgres/qgspgnewconnection.cpp]:
```suggestion
    cb_projectsInDatabase->setToolTip( tr( "Saving projects in databases not available for Postgres databases earlier than 9.5" ) );
```

File: src/providers/postgres/qgspgnewconnection.cpp [hunk 4]
```diff
@@ -205,6 +206,18 @@ void QgsPgNewConnection::testConnection()
 
   QgsPostgresConn *conn = QgsPostgresConn::connectDb( uri.connectionInfo( false ), true );
 
+  if ( conn->pgVersion() < 90500 )
+  {
+    cb_projectsInDatabase->setChecked( false );
```
Comment (m-kuhn) [inline src/providers/postgres/qgspgnewconnection.cpp]: What's this for?

**Comment by gioman** [comment]
@stev-0 
if the goal of this patch was to disable saving projects within PostgreSQL DBs if postgresql is < 9.5 then it does not work on both LTR and 3.12. Users are still able to try save the project and get an error message in return.

**Comment by stev-0** [comment]
@gioman possibly this is an old connection that was previously setup? Any new connection that is created should disallow save as project, but this doesn't work retrospectively, if I remember rightly because it would be expensive to check for this apart from at the moment the connection is first created, but there may be another way to deal with this, if this is indeed the issue.

**Comment by gioman** [comment]
> possibly this is an old connection that was previously setup?

@stev-0 yes, but it happens also with newly created connections.

**Pull request merged**